sql/core/src/main/scala/org/apache/spark/sql/execution/DataSourceScanExec.scala

```diff
@@ -306,7 +306,15 @@ case class FileSourceScanExec(
       withOptPartitionCount
     }
 
-    withSelectedBucketsCount
+    val withOptColumnCount = relation.fileFormat match {
+      case columnar: ColumnarFileFormat =>
+        val sqlConf = relation.sparkSession.sessionState.conf
+        val columnCount = columnar.columnCountForSchema(sqlConf, requiredSchema)
+        withSelectedBucketsCount + ("ColumnCount" -> columnCount.toString)
```

Comment:
Can't we know column count from `requiredSchema` metadata?

Reply:
You can "guess-timate" the physical column count by counting the leaf fields in the `ReadSchema` metadata value, but the true answer is an implementation issue of the file format. For example, in the implementation of `ColumnarFileFormat` for Parquet, we convert the the Catalyst schema to the Parquet schema before counting columns. I suppose a similar approach would be required for ORC and other columnar formats.

That being said, this new metadata value isn't really meant to provide new and essential information, _per se_. Its purpose is to provide easy-to-read, practical information that's useful for quickly validating that schema pruning is working as expected. For example, seeing that a query is reading all 423 columns from a table instead of 15 tells us pretty quickly that schema pruning is not working (unless we really are trying to read the entire table schema). I've found the `ReadSchema` value to be difficult to read in practice because of its terse syntax, and because its printout is truncated.

Reply:
shall we only include this info when the columnar reader is on?

Reply:
Is this something we really should include in the metadata? If the purpose of this is to check if the column pruning works or not, logging should be good enough. Adding a trait for it sounds an overkill for the current status. Let's not add an abstraction just for rough guess that it can be generalised.

Reply:
The purpose of this info is to check the number of columns actually selected, and that information can be shown via logging, no? Why should it be exposed in metadata then?

Maybe debug logging that shows the number of columns that actually being selected via the underlying source.

Reply:
> The purpose of this info is to check the number of columns actually selected, and that information can be shown via logging, no? Why should it be exposed in metadata then?

Yes, this information can be shown in logging, but I would say that getting access to the right log and identifying the right message is quite a bit more effort. Consider also the case of the beeline user connecting to a multiuser thriftserver. They are pretty far from the log file, whereas running an 'explain' is right there in the terminal.

I think setting a log level to get this information is a bit overkill. You're going to opt in to a lot of information you don't want.

Reply:
I was wondering how important to know if the columns are pruned or not. In that way, other logs should be put in metadata. For instance, we're not even showing the actual filters (not cayalyst but I mean the actual pushed filters that are going to apply at each source implementation level such as filters from `ParquetFilters.createFilter`) in Spark side.

Reply:
> In that way, all other logs should be put in metadata.

Who wants that? If someone wants to put metadata somewhere in the physical plan, let them open a PR and make a case for it. Otherwise, why would we put things into the plan metadata that no one has asked for? That does not make sense.

> For instance, we're not even showing the actual filters (not cayalyst but I mean the actual pushed filters that are going to apply at each source implementation level such as filters from ParquetFilters.createFilter) in Spark side.

That's right, however that's not the aim of this PR. The aim is to ensure that plan-time schema pruning is working as expected, and I assert that that information itself is valuable enough to warrant inclusion in the data source metadata. That's speaking from experience, not conjecture. This metadata has been a _very valuable_ tool for us, and incurs little burden in practice.

Reply:
> Who wants that? If someone wants to put metadata somewhere in the physical plan, let them open a PR and make a case for it

No .. I don't think we should add it only because it's requested once. They look same instances to me. I will have no argument if this one is added and other people request to add others later. We should make it clear why this one should be specifically added. We're not going to add all the information to metadata as requested.

If the purpose of adding it is to check if the pushing down is actually working or not, the logging sounds appropriate for its purpose.

Reply:
1. The column pruning is now specific for Parquet .. it's source specific for now. 2. I really think it's more appropriate to check if something as expected or not by logging.

> That's speaking from experience, not conjecture.

I am not underestimating your statement. Let's be very clear why it should be put in metadata over logging. How and why it can be useful? in what cases?

Reply:
I'll reiterate a sample use case:

> Consider also the case of the beeline user connecting to a multiuser thriftserver. They are pretty far from the log file, whereas running an 'explain' is right there in the terminal.

This also matters to users planning/debugging queries in a Jupyter notebook, as we have in VideoAmp. The LOE for these users to go to a driver log file is quite high by comparison to inspecting a query plan.

When you refer to logging, which log are you referring to? When would this information be logged? And at what log level?

Reply:
That basically says logging is useless when to use beeline. I don't think this info is super important to (non-advanced) users.

I mean log4j which is Spark's logging module, and I meant information you're including in the metadata. Maybe info level? or debug level.

```
logInfo("The number of actual column being pruned is blah blah")
```

Reply:
> That basically says logging is useless when to use beeline. I don't think this info is super important to (non-advanced) users.

My experience says otherwise, and advanced users use beeline and Jupyter, too.